{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyOS/NeojgDSfydnbd9Q+beR",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/recanoy/Coursera_AlgorithmsForDNASequencing/blob/main/Week2_LectureNotes.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Author: Raymart Jay E. Canoy\n",
        "# Date: 16 October 2022\n",
        "___\n",
        "\n",
        "* **Outline**\n",
        "  1. **Boyer-Moore Algorithm**: An alternative to naive algorithm\n",
        "  2. **Indexing**: Working on genomes and read-alignment problem\n",
        "  3. **Pigeon-hole Principle**: On exact and approximare matching problems\n"
      ],
      "metadata": {
        "id": "zDnr0kBgcmIg"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### (1) Boyer-Moore basics\n",
        "\n",
        "* Learn from character comparisons to skip pointless alignments\n",
        "* Try alignments in left-to-right order, and try character comparisons in right-to-left order\n",
        "* **Bad character rule:**\n",
        "  Upon mismatch, skip alignments until\n",
        "  (a) mismatch becomes a match, or\n",
        "  (b) P moves past mismatched character\n",
        "* **Good suffix rule:**\n",
        "  Let $t=$ substring matched by inner loop; skip until\n",
        "  (a) there are no mismatches between $P$ and $t$ or\n",
        "  (b) $P$ moves past $t$"
      ],
      "metadata": {
        "id": "Y6mYjZE7cnBp"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Diversion: Repetitive elements\n",
        "\n",
        "* Transposable elements: Infiltrates the genome\n",
        "* 45% in the human genome came from transposable elements\n",
        "  * For example, `Alu` (11%)\n",
        "  * Read: Cordaux R, Batzer MA. The impact of retrotransposons on human genome evolution. Nat Rev Genet. 2009 Oct; 10(10):691-703\n",
        "* **Relevance of transposable elements**\n",
        "  * Transposable elements create ambiguity in the human genome\n",
        "  * Repettive sequence in the genome are gonna create problem in our problem."
      ],
      "metadata": {
        "id": "cIHUzEX86Kh4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import string\n",
        "\n",
        "def z_array(s):\n",
        "  \"\"\"\n",
        "  Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s\n",
        "  \"\"\"\n",
        "  z = [len(s)] + [0] * (len(s)-1)\n",
        "\n",
        "  # Initial comparison of s[1:] with prefix\n",
        "  for i in range(1, len(s)):\n",
        "    if s[i] == s[i-1]:              # compares the nucleotide in the previos index matches with the current nucleotide\n",
        "      z[1] += 1                     # if yes, add 1 to the the z list at index 1\n",
        "    else:                           # otherwise, break\n",
        "      break\n",
        "\n",
        "  r, l = 0, 0\n",
        "  if z[1] > 0:\n",
        "    r, l = z[1], 1\n",
        "  \n",
        "  for k in range(2, len(s)):\n",
        "    assert z[k] == 0\n",
        "    if k > r:\n",
        "      # Case 1\n",
        "      for i in range(k, len(s)):\n",
        "        if s[i] == s[i-k]:\n",
        "          z[k] += 1\n",
        "        else:\n",
        "          break\n",
        "      r, l = k + z[k] - 1, k\n",
        "    else:\n",
        "      # Case 2\n",
        "      # Calculate length of beta\n",
        "      nbeta = r - k + 1\n",
        "      zkp = z[k-l]\n",
        "      if nbeta > zkp:\n",
        "        # Case 2a: Zkp wins\n",
        "        z[k] = zkp\n",
        "      else:\n",
        "        # Case 2b: Compare characters just past r\n",
        "        nmatch = 0\n",
        "        for i in range(r+l, len(s)):\n",
        "          if s[i] == s[i-k]:\n",
        "            nmatach += 1\n",
        "          else:\n",
        "            break\n",
        "        l, r = k, r + nmatch\n",
        "        z[k] = r - k + 1\n",
        "  \n",
        "  return z\n",
        "\n",
        "\n",
        "def n_array(s):\n",
        "  \"\"\"\n",
        "  Compare the N array (Gusfield theorem 2.2.2) from the Z array\n",
        "  \"\"\"\n",
        "  return z_array(s[::-1])[::-1]\n",
        "\n",
        "def big_l_prime_array(p, n):\n",
        "  \"\"\"\n",
        "  Compile L' array (Gusfield theorm 2.2.2) using p and N array.\n",
        "  L'[i] = largest index j less than n such that N[j] = |P[i:]|\n",
        "  \"\"\"\n",
        "  lp = [0] * len(p)\n",
        "  for j in range(len(p)-1):\n",
        "    i = len(p) - n[j]\n",
        "    if i < len(p):\n",
        "      lp[i] = j + 1\n",
        "\n",
        "  return lp\n",
        "\n",
        "def big_l_array(p, lp):\n",
        "  \"\"\"\n",
        "  Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
        "  L[i] = largest index j less than n such that N[j] >= |P[i:]|\n",
        "  \"\"\"\n",
        "  l = [0] * len(p)\n",
        "  l[1] = lp[1]\n",
        "  for i in range(2, len(p)):\n",
        "    l[i] = max([i-1], lp[i])\n",
        "  \n",
        "  return l\n",
        "\n",
        "def small_l_prime_array(n):\n",
        "  \"\"\"\n",
        "  Compile lp' array (Gusfield theorem 2.2.4) using N array.\n",
        "  \"\"\"\n",
        "  small_lp = [0] * len(n)\n",
        "  for i in range(len(n)):\n",
        "    if n[i] == i+1:                 # prefix matching a suffix\n",
        "      small_lp[len(n)-i-1] = i+1\n",
        "  for i in range(len(n)-2, -1, -1): # \"smear\" them out to the left\n",
        "    if small_lp[i] == 0:\n",
        "      small_lp[i] = small_lp[i+1]\n",
        "  \n",
        "  return small_lp\n",
        "\n",
        "def good_suffix_table(p):\n",
        "  \"\"\"\n",
        "  Return tables needed to apply good suffix rule.\n",
        "  \"\"\"\n",
        "  n = n_array(p)\n",
        "  lp = big_l_prime_array(p, n)\n",
        "  return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
        "\n",
        "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
        "  \"\"\"\n",
        "  Given a mismatch at offset i, and given L/L' and l' arrays,\n",
        "  return amount to shift as determined by good suffix rule.\n",
        "  \"\"\"\n",
        "  length = len(big_l_prime)\n",
        "  assert i < length\n",
        "  if i == length - 1:\n",
        "    return 0\n",
        "  i += 1      # i points to leftmost matchig position of p\n",
        "  if big_l_prime[i] > 0:\n",
        "    return length - big_l_prime[i]\n",
        "  return length - small_l_prime[i]\n",
        "\n",
        "def good_suffix_match(small_l_prime):\n",
        "  \"\"\"\n",
        "  Given a full match of P to T, return amount to shift as\n",
        "  determined by good suffix rule.\n",
        "  \"\"\"\n",
        "  return len(small_l_prime) - small_l_prime[1]\n",
        "\n",
        "def dense_bad_char_tab(p, amap):\n",
        "  \"\"\"\n",
        "  Given pattern string and list with ordered alphabet characters, create\n",
        "  and return a dense bad character table. Table is indexed by offset\n",
        "  then by character.\n",
        "  \"\"\"\n",
        "  tab = []\n",
        "  nxt = [0] * len(amap)\n",
        "  for i in range(0, len(p)):\n",
        "    c = p[i]\n",
        "    assert c in amap\n",
        "    tab.append(nxt[:])\n",
        "    nxt[amap[c]] = i + 1\n",
        "  return tab"
      ],
      "metadata": {
        "id": "EaLRGwvZ1kAv"
      },
      "execution_count": 13,
      "outputs": []
    }
  ]
}